src/traces/table/plot.js

```diff
@@ -0,0 +1,892 @@
+
+    cellText.enter()
+        .append('text')
+        .classed('cellText', true);
+
+    return cellText;
+}
+
+function renderCellTextHolder(columnCell) {
+
+    var cellTextHolder = columnCell.selectAll('.cellTextHolder')
+        .data(gup.repeat, function(d) {return d.keyWithinBlock;});
+
+    cellTextHolder.enter()
+        .append('g')
+        .classed('cellTextHolder', true);
+
+    return cellTextHolder;
+}
+
+function setFont(columnCell) {
+    columnCell
+        .each(function(d, i) {
+            var spec = d.calcdata.cells.font;
+            var col = d.column.specIndex;
+            var font = {
+                size: gridPick(spec.size, col, i),
+                color: gridPick(spec.color, col, i),
+                family: gridPick(spec.family, col, i)
+            };
+            Drawing.font(d3.select(this), font);
+
+            d.rowNumber = d.key;
+            d.align = gridPick(d.calcdata.cells.align, col, i);
+            d.valign = gridPick(d.calcdata.cells.valign, col, i);
+            d.cellBorderWidth = gridPick(d.calcdata.cells.line.width, col, i);
+            d.font = font;
+        });
+}
+
+function sizeAndStyleRect(cellRect) {
+    cellRect
+        .attr('width', function(d) {return d.column.columnWidth;})
+        .attr('stroke-width', function(d) {return d.cellBorderWidth;})
+        .attr('stroke', function(d) {
+            return c.clipView ?
+                ({header: 'blue', cells1: 'red', cells2: 'green'})[d.column.key] :
+                gridPick(d.calcdata.cells.line.color, d.column.specIndex, d.rowNumber);
+        })
+        .attr('fill', function(d) {
+            return d.calcdata.cells.fill ? gridPick(d.calcdata.cells.fill.color, d.column.specIndex, d.rowNumber) : 'none';
+        });
+}
+
+function populateCellText(cellText, tableControlView, allColumnBlock, gd) {
+    cellText
+        .text(function(d) {
+            var col = d.column.specIndex;
+            var row = d.rowNumber;
+            var userSuppliedContent = d.value;
+            var latex = isLatex(userSuppliedContent);
+            var userBrokenText = (typeof userSuppliedContent !== 'string') || userSuppliedContent.match(/<br>/i);
+            var userBrokenText2 = (typeof userSuppliedContent === 'string') && userSuppliedContent.match(/<br>/i);
+            var prefix = latex ? '' : gridPick(d.calcdata.cells.prefix, col, row) || '';
+            var suffix = latex ? '' : gridPick(d.calcdata.cells.suffix, col, row) || '';
+            var format = latex ? null : gridPick(d.calcdata.cells.format, col, row) || null;
+            var prefixSuffixedText = prefix + (format ? d3.format(format)(d.value) : d.value) + suffix;
+            d.latex = latex;
+            d.mayHaveMarkup = (typeof userSuppliedContent === 'string') && userSuppliedContent.match(/[<&>]/);
+            var hasWrapSplitCharacter;
+            var hwsc = function(prefixSuffixedText) {return prefixSuffixedText.indexOf(c.wrapSplitCharacter) !== -1;};
```

Comment:
Oh boy. I really wished the text-wrapping logic would've been added to `svg_text_utils.js` so that other parts of plotly.js can use it.

@monfera is there any reason to keep this here?

Reply:
Agreed, it should be extracted out. I scrambled during the implementation, as the significant code size of `svg_text_utils`, the previously unknown (to me) MathJax, the need for doing asynchronous cell content population and bumping row heights and revolver panel offsets on taller content yielded quite a few challenges. 

Maybe I'd need to consider alternative work processes - and glad to receive input on this - but it felt more manageable to work out the kinks while not risking other plot types, and then unify them once it proved itself. You were kind to offer developing some parts of `table` and I think we even talked about specific items (scrolling etc.), maybe my avoidance of bothering you w/ these wasn't ideal.

Reply:
> I really wished the text-wrapping logic would've been added to svg_text_utils.js so that other parts of plotly.js can use it.

That would certainly be the great, and it will have to happen at some point. I feel like a lot of the complexity here has to do with:
- figuring out whether we're allowed to apply formatting or not - which, you're right, doesn't belong in `svg_text_utils`, but I feel like it would be simpler and more robust to *try* to apply formatting if the user has specified formatting, then back off (and just cast as string?) if it didn't work, rather than trying to determine ahead of time whether formatting is allowed. Then perhaps we wouldn't need all the `typeof` checks and the confusing `userBrokenText` and `userBrokenText2`?
- trying to figure out if we can avoid `convertToTspans` altogether in favor of [this block](https://github.com/plotly/plotly.js/pull/2052/files?diff=unified#diff-1e0dea3a9361823fe80cbbdc50befd7eR509), I guess for performance purposes? But if shortcuts like that exist to improve over the performance of `convertToTspans`, shouldn't we use those shortcuts *in* `convertToTspans` so everyone can benefit?

If this section *does* need to have this kind of logic, I feel like it could be made a lot clearer with `if` branches instead of ternary operators.

Reply:
@alexcjohnson yes the naming of certain vars in the determination of which code paths to hit would benefit from improvement; I still wanted to spend a bit of time cleaning up that section, also, logic-wise (and we could save some CPU cycles, some `regex` is called twice).

`convertToTspans` is long code that does much heavy lifting and my perf measurements showed it to be a main bottleneck, which is why it's avoided when possible (this code split was a reaction to benchmarking). I think that most often, users don't have a simultaneous need for cell content formatting _and_ actual infinity scrolling with no jank. So, as you can see in the Bird example, scrolling is pretty smooth with plain content (we have more headroom if need be) even though it applies `D3` numerical formatting, pre/postfix, alignment.

I agree with the suggestion that such performance-minded shortcuts may benefit all callers of `convertToTspans` and it boils down to two things:

1. Development time, which is dominated by having to ensure, also with manual testing, that nothing breaks; also, getting more intimate in `convertToTspan` as it's also imperative code that needs time (I found that doing two hard things at the same time is non-linear, risk-wise and I was already at my comfort level). 
2. Is there a current or foreseen need? Saving CPU cycles is always a good idea, though in a lot of cases, plots have but a handful or maybe a couple of dozen text elements, and even if more, often, there's no need to support their redrawing with a high frame rate as in a scroller. `table` is quite special in that it's both absolutely text heavy and benefits from very fast renders, otherwise scroll is janky.

One more note about `convertToTspan`, which generalizes to our other functions such as `Drawing.font` (already in use in `table`) and other things that @etpinard suggested in this review such as `dragElement`, other scrollers etc. Basically all these things couple computational tasks with rendering tasks. Moreover, rendering is in-place and often singleton, ie. requires a `gd` object, builds on a number of assumptions which were naturally true when making them or extracting them out. It sounds abstract so here's an example:

LaTeX rendering has high latency. I started coding toward saving previously shown cells so that, if the user has already scrolled there, there would be no need for a fresh call to `convertToTspan`. But the way it works is, it _computes_ what to put in the DOM and then _renders_ it too. So I can't just cache the results and reapply it anytime. Faced with the choice of querying and saving DOM fragments, and subsequently reapplying them, I thought let's find some other way. Maybe the best place to do caching is in `convertToTspan` itself but this component may not have enough info on when to release (memory leak). What would yield more flexibility is,

1. the component would return _something_
2. that _something_ could be manipulated by the caller who has info on memory use, eg. it can bind memory to DOM elements which go away after not needed (GC) or the trace can use some MRU/LRU cache component as it befits

The _something_ could be 

- a HTML `innerText`/`outerText` (obv. limited as it can't do just styling like mixing color)
- a function that takes a DOM node and alters or recreates desired structures
- a virtual DOM fragment

I haven't thought much about it but some small vDOM utility like Preact (3kB) would do the job (adding React would be too much payload for a task like this and Preact API is almost identical).

In general, I wouldn't mind eventually breaking up these utilities and uncoupling calculations (-> pure functions) and rendering (which, if we want, can also be represented declaratively, eg. via some `diff` or `ensure` patch not unlike a vDOM object, or if not, it can be a returned function that takes a DOM node as input).

Reply:
> Development time, which is dominated by having to ensure, also with manual testing, that nothing breaks; also, getting more intimate in `convertToTspan` as it's also imperative code that needs time (I found that doing two hard things at the same time is non-linear, risk-wise and I was already at my comfort level).

That's fair, I know `convertToTspans` is a cumbersome chunk of code even after several attempts to clean it up over the years. We may also want to order our pushes wisely - for example perhaps we explore KaTeX first, then merge wrapping in. Please make an issue to tackle this though. There shouldn't be too much manual testing needed, between images and jasmine we have fairly good coverage of `convertToTspans`

> Is there a current or foreseen need?

The biggest need is for functionality (wrapping) rather than perf, though the latter certainly wouldn't hurt. You're right that table is the heaviest user of text, but there are a couple of cases that can get pretty heavy during dynamic zoom/pan:
- tick labels, particularly categories
- scatter text labels

> In general, I wouldn't mind eventually breaking up these utilities and uncoupling calculations (-> pure functions) and rendering

That sounds like a good general strategy, my concern is mainly about a proliferation of different patterns scattered about the codebase, with no clear plan to reunify them later. That's the reason for the fairly strong preference to use what's present and either fix or extend it to fit new needs, rather than augmenting it from the outside.

Reply:
Added the requested separate issue: https://github.com/plotly/plotly.js/issues/2053